File: cpp/2309_greatest_letter/greatest_letter.cpp
/*
 * @Date: 2023-01-27
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-01-27
 * @FilePath: /algorithm/cpp/2309_greatest_letter/greatest_letter.cpp
 */

#include <cassert>
#include <string>

using namespace std;

class Solution {
 public:
  string greatestLetter(string s) {
    int lower = 0, upper = 0;
    for (char& c : s) {
      if (islower(c)) {
        lower |= 1 << (c - 'a');
      } else {
        upper |= 1 << (c - 'A');
      }
    }
    for (int i = 25; i >= 0; i--) {
      if (lower & upper & (1 << i)) {
        return string(1, 'A' + i);
      }
    }
    return "";
  }
};

int main() {
  {
    string s = "lEeTcOdE";
    string ans = "E";
    assert(Solution().greatestLetter(s) == ans);
  }

  {
    string s = "arRAzFif";
    string ans = "R";
    assert(Solution().greatestLetter(s) == ans);
  }

  {
    string s = "AbCdEfGhIjK";
    string ans = "";
    assert(Solution().greatestLetter(s) == ans);
  }
}

File: golang/2309_greatest_letter/greatest_letter.go
/*
 * @Date: 2023-01-27
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-01-27
 * @FilePath: /algorithm/golang/2309_greatest_letter/greatest_letter.go
 */

package main

import (
	"math/bits"
	"unicode"
)

func greatestLetter(s string) string {
	mask1, mask2 := 0, 0
	for _, c := range s {
		if unicode.IsLower(c) {
			mask1 |= 1 << (c - 'a')
		} else {
			mask2 |= 1 << (c - 'A')
		}
	}
	mask := mask1 & mask2
	if mask == 0 {
		return ""
	}
	return string(byte(bits.Len(uint(mask))-1) + 'A')
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}

	{
		s := "lEeTcOdE"
		ans := "E"
		assert(greatestLetter(s) == ans)
	}

	{
		s := "arRAzFif"
		ans := "R"
		assert(greatestLetter(s) == ans)
	}

	{
		s := "AbCdEfGhIjK"
		ans := ""
		assert(greatestLetter(s) == ans)
	}
}
